[Native clouds] dependency injection for Azure resources

istraka opened the pull request:
FYI @mchoma

mnovak1 reviewed (approved):
LGTM

mnovak1 commented:
@istraka  I'll not merge it in case that anyone else wants to review as well.

mchoma commented on clouds/pom.xml:
```diff
@@ -78,6 +78,7 @@
                         <excludes>
                             <!-- These should not be run alone since they are supposed to be run in a suite -->
                             <exclude>**/suite/tests/*.java</exclude>
+                            <exclude>**/suitetests/*.java</exclude>
```
is this expected? I do not see such folder.

istraka replied:
It is in `azure.armTemplates.di.suitetests`. It is there to exclude those test classes so that they are note executed alone but only suite including them is executed.

mchoma replied:
Ok

mchoma commented on clouds/clouds-azure/src/test/java/azure/armTemplates/AzureTestConstants.java:
```diff
@@ -6,4 +6,8 @@ public class AzureTestConstants {
 
     public static final String VNET_NAME_2 = "sunstoneVnet2";
+
+    public static final String IMAGE_REF = "${image.ref:/subscriptions/7dee6f21-9f05-414e-99fa-08d3215fb420/resourceGroups/istraka-test/providers/Microsoft.Compute/images/eap-test-image}";
```
that default is allright?

istraka replied:
This one is OK for now. It is my test image, and I don't have a strategy how to handle it, yet.

mchoma replied:
strategy: something released as default (or first version which will be released) and development image provided as parameter?

istraka replied:
Yeah, something like that. Or some test image living in our subscription.

mchoma commented on clouds/clouds-azure/src/test/java/azure/armTemplates/di/SunstoneResourceAzSuiteTests.java:
```diff
@@ -0,0 +1,13 @@
+@Suite
+@SelectClasses({AzEapHostnameTests.class, AzEapManagementClientTests.class})
+public class SunstoneResourceAzSuiteTests {
+    public static final String group = "${az.group:istraka}";
```
`istraka` default is ok?

istraka replied:
I have changed it.

mchoma commented on clouds/clouds-azure/src/test/resources/azure/armTemplates/eap.json:
```diff
@@ -0,0 +1,182 @@
+    },
+    "adminUsername": {
+      "type": "string",
+      "defaultValue": "istraka"
```
some service account here?

istraka replied:
No, a user for eap management. I will change it to sunstone-something.

istraka replied:
Done.

mchoma commented on clouds/clouds-azure/src/test/resources/azure/armTemplates/eap.json:
```diff
@@ -0,0 +1,182 @@
+    },
+    "imageRefId": {
+      "type": "string",
+      "defaultValue": "/subscriptions/7dee6f21-9f05-414e-99fa-08d3215fb420/resourceGroups/istraka-test/providers/Microsoft.Compute/images/eap-test-image"
```
replace with `AzureTestConstants.IMAGE_REF` ?

istraka replied:
I removed the default value. It still comes from test as a parameter.

mchoma commented on clouds/clouds-azure/src/test/java/azure/armTemplates/di/suitetests/AzEapHostnameTests.java:
```diff
@@ -0,0 +1,49 @@
+        template = "azure/armTemplates/eap.json", group = group, perSuite = true)
+public class AzEapHostnameTests {
+
+    @AzureVirtualMachine(name = instanceName)
```
Could you proide default value for `name` parameter  value `azure.armTemplates.AzureTestConstants.instanceName`? So user can write  just

```
  @AzureVirtualMachine
  Hostname eapHostname;
```

mchoma replied:
We have talked this is not possible because instance name must be taken from Azure template.

mchoma commented on clouds/clouds-azure/src/main/java/azure/core/identification/AzureInjectJavaDoc.java:
```diff
@@ -0,0 +1,53 @@
+ *     </tr>
+ * </table>
+ */
+public interface AzureInjectJavaDoc {
```
Interesting idea :)

istraka replied:
Removed

mchoma commented:
I run through and I see it is aligned with what we have discussed. I have left just some minor comments.

mnovak1 commented on clouds/clouds-azure/src/main/java/azure/core/AzureIdentifiableSunstoneResource.java:
```diff
@@ -0,0 +1,152 @@
+        @Override
+        <T> T getResource(Annotation injectionAnnotation, AzureSunstoneStore store, Class<T> clazz) throws SunstoneException {
+            if(!AzureVirtualMachine.class.isAssignableFrom(injectionAnnotation.annotationType())) {
+                throw new IllegalArgumentSunstoneException("todo");
```
Exception message should be updated.

istraka replied:
Fixed.

mnovak1 commented on clouds/clouds-azure/src/main/java/azure/core/AzureIdentifiableSunstoneResource.java:
```diff
@@ -0,0 +1,152 @@
+        @Override
+        <T> T getResource(Annotation injectionAnnotation, AzureSunstoneStore store, Class<T> clazz) throws SunstoneException {
+            if(!AzureVirtualMachine.class.isAssignableFrom(injectionAnnotation.annotationType())) {
+                throw new IllegalArgumentSunstoneException("todo");
```
Same here.

istraka replied:
Fixed.

istraka commented:
I have fixed issues and changed the SPI a little - core module does the injection and SPI services are for creating an object to be injected based on Annotation and the type.

istraka commented:
Ready.

istraka commented on clouds/clouds-azure/src/main/java/azure/core/AzureIdentifiableSunstoneResource.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.Arrays;
+
+/**
+ * Enum of AWS resources that can be identified by {@link SunstoneResource#value()} and what fild types are supported
```
Fixed.
